expo/MCTS.py

```diff
@@ -0,0 +1,457 @@
+            mcts_logger.log("MCTS", f"Number of simulations: {self.get_num_simulations()}")
+            mcts_logger.log("MCTS", f"Tree loaded: {tree_loaded}")
+
+        if not tree_loaded:
+            rollouts -= 2  # 2 rollouts for the initial tree
+            if rollouts < 0:
+                raise ValueError("Rollouts must be greater than 2 if there is no tree to load")
+            self.children[root] = []
+            reward = await self.simulate(root, role)
+            self.backpropagate(root, reward)
+            node, reward = await self.expand_and_simulate(root)
+            # self.backpropagate(node, reward)
+            self.save_node_order(root.id)
+            self.save_node_order(node.id)
+        else:
+            root = self.root_node
+            self.load_node_order()
+
+        for _ in range(rollouts):  # number of rollouts
+            mcts_logger.log("MCTS", f"Start the next rollout {_+1}")
+            node = self.select(root)
+            if node.is_terminal():
+                if node.raw_value == 0:
+                    reward = await self.simulate(node)
+                else:
+                    reward = {"test_score": node.raw_value, "score": node.raw_reward["score"]}
+                mcts_logger.log("MCTS", f"Terminal node's reward: {reward}")
+                self.backpropagate(node, reward)
+            else:
+                node, reward = await self.expand_and_simulate(node)
+                # self.backpropagate(node, reward)
+            self.save_node_order(node.id)
+        return self.best_path(root)
+
+    async def expand_and_simulate(self, node):
+        # Expand and randomly select a child node, then simulate it
+        if node.visited > 0:
+            children = await self.expand(node)
+            node = random.choice(children)
+        reward = await self.simulate(node)
+        self.backpropagate(node, reward)
+        return node, reward
+
+    def load_tree(self):
+        def load_children_node(node):
+            mcts_logger.log("MCTS", f"Load node {node.id}'s child: {node.children}")
+            if node.is_terminal() or not node.children:
+                return
+            for child in node.children:
+                child.load_node()
+                self.children[child] = child.children
+                load_children_node(child)
+
+        # Load all pkl files in the node_dir
+        all_pkl_files = os.listdir(self.root_node.state["node_dir"])
+        all_pkl_files = [f for f in all_pkl_files if f.endswith(".pkl")]
+        if os.path.exists(os.path.join(self.root_node.state["node_dir"], "Node-0.pkl")):
+            with open(os.path.join(self.root_node.state["node_dir"], "Node-0.pkl"), "rb") as f:
+                self.root_node = pickle.load(f)
+            self.children[self.root_node] = self.root_node.children
+            load_children_node(self.root_node)
+
+            if self.children:
+                return True
+        return False
```

Comment:
Run pre-commit, specifically, use https://github.com/geekan/MetaGPT/blob/main/.pre-commit-config.yaml to format the code